File: 프로그래머스/3/152995. 인사고과/인사고과.swift
import Foundation

func solution(_ scores: [[Int]]) -> Int {
    var answer = 0
    var indexedScores = scores.enumerated().map { (i, score) in return [i, score[0], score[1], 0] }
    indexedScores.sort {
        if $0[1] != $1[1] { // 내림 차순
            return $0[1] > $1[1]
        } else {
            return $0[2] < $1[2] // 오름 차순
        }
    }

    var disqualified = Set<Int>()
    var maxB = -1
    for score in indexedScores {
        if score[2] < maxB {
            disqualified.insert(score[0])
        } else {
            maxB = max(maxB, score[2])
        }
    }

    if disqualified.contains(0) {
        return -1
    }

    var rankList = indexedScores.filter { !disqualified.contains($0[0]) }
    rankList.sort { // 점수 정렬
        $0[1] + $0[2] > $1[1] + $1[2]
    }

    rankList[0][3] = 1
    for i in 1..<rankList.count {
        if rankList[i][1] + rankList[i][2] == rankList[i-1][1] + rankList[i-1][2] {
            rankList[i][3] = rankList[i-1][3]
        } else {
            rankList[i][3] = i + 1
        }
    }

    for score in rankList {
        if score[0] == 0 {
            answer = score[3]
            break
        }
    }

    return answer
}

File: 프로그래머스/3/152995. 인사고과/인사고과.py
def solution(scores):
    answer = 0
    scores = [[i, score[0], score[1], 0] for i, score in enumerate(scores)]
    scores.sort(key=lambda x: (-x[1], x[2]))  # 점수1 내림차순, 점수2 오름차순

    tmp = set()
    max_b = -1
    for i in range(len(scores)):
        if scores[i][2] < max_b:
            tmp.add(scores[i][0])
        else:
            max_b = scores[i][2]

    if 0 in tmp:
        return -1
    rank = [score for score in scores if score[0] not in tmp]
    rank.sort(reverse=True, key= lambda x: x[1]+x[2])
    rank[0][3] = 1
    for i in range(1, len(rank)):
        if rank[i][1] + rank[i][2] == rank[i-1][1] + rank[i-1][2]:
            rank[i][3] = rank[i-1][3]
        else:
            rank[i][3] = i+1
    for i in range(len(rank)):
        if rank[i][0] == 0:
            answer = rank[i][3]
            break
    return answer 